R/S3.R

```diff
@@ -182,16 +189,64 @@ validate_data.frame <- function(self) {
   }
 }
 
+valid_dimnames <- function(self) {
+  dn <- dimnames(self)
+  if (is.null(dn))
+    TRUE
+  else if (!is.list(dn) || length(dn) != length(dim(self)))
+    FALSE
+  else for (i in seq_along(dimnames(self))) {
+    if (is.null(dn[[i]]))
+      next
+    if (!is.character(dn[[i]]) || length(dn[[i]]) != dim(self)[[i]])
+      return(FALSE)
+  }
+  TRUE
+}
+
+validate_matrix <- function(self) {
+  if (is.matrix(self)) # is.matrix() methods should only return TRUE if valid
+    return(invisible(NULL))
+  if (!is.integer(dim(self)) || length(dim(self)) != 2L || !all(dim(self) >= 0L))
+    return("dim(self) must be a non-negative integer vector of length 2")
+  if (!valid_dimnames(self))
+    return("dimnames(self) must be NULL or a length 2 list of either NULL or a character vector of length equal to its corresponding dimension")
+  "is.matrix(self) is FALSE"
```

Comment:
What do you think of this style?

```suggestion
  if (!is.matrix(self)) {
    # is.matrix() methods should only return TRUE if valid
    "is.matrix(self) is FALSE"
  } else if (!is.integer(dim(self)) || length(dim(self)) != 2L || !all(dim(self) >= 0L)) {
    "dim(self) must be a non-negative integer vector of length 2"
  } else if (!valid_dimnames(self)) {
    "dimnames(self) must be NULL or a length 2 list of either NULL or a character vector of length equal to its corresponding dimension"
  }
```